flang/lib/Lower/Bridge.cpp

```diff
@@ -403,18 +403,21 @@ class FirConverter : public Fortran::lower::AbstractConverter {
               [&](Fortran::lower::pft::FunctionLikeUnit &f) {
                 if (f.isMainProgram())
                   hasMainProgram = true;
-                declareFunction(f);
+                createGlobalOutsideOfFunctionLowering(
```

Comment:
Could you make that a single call that wraps the whole `for (Fortran::lower::pft::Program::Units &u` ...`, there is no real need to create an delete a  mock builder for each declaration.

I think that would involve modifying `createGlobalOutsideOfFunctionLowering` to check if a builder is already set and just directly call the callback in that case to allow nested call.

Also this should probably be renamed to something like `createBuilderOutsideOfFuncOpAndDo` if used for other things than global creations.